modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java

```diff
@@ -85,7 +88,19 @@ public ExecutionPlan map(MappingService mappingService, MappingQueryContext ctx)
                 else
                     ex.addSuppressed(e);
 
-                fragments = replace(fragments, e.fragment(), new FragmentSplitter(e.node()).go(e.fragment()));
+                RelNode cutPoint = e.node();
+
+                // TableModify inside transaction must be executed locally.
+                boolean forceLocTableModify = Commons.queryTransactionVersion(ctx) != null && cutPoint instanceof IgniteTableModify;
+
+                if (forceLocTableModify)
+                    cutPoint = ((SingleRel)cutPoint).getInput(); // Cuts TableScan instead of TableModification.
```

Comment:
After this split we have:
fragment 1 with TableModify and broadcast distribution 
fragment 2 with sender to broadcast distribution
It's not correct, since second fragment will send data to all nodes and there are some hacks required to change TableModify (mapLocalTableModify)

Reply:
Forced single distribution for IgniteReceiver.